File: Universe_Backend/Data/Models/Message.cs
using Universe_Backend.Data.DTOs;

namespace Universe_Backend.Data.Models
{
    public class Message
    {
        public required int Id { get; set; }
        public required string Body { get; set; }
        public List<string> Images { get; set; } = [];
        public List<string> Videos { get; set; } = [];
        public List<string> Audios { get; set; } = [];
        public DateTime PublishDate { get; set; } = DateTime.Now;
        public int? ChildPostId { get; set; }
        public int? MessageRepliedTo { get; set; }
        public required string AuthorId { get; set; }
        public ICollection<Widget>? Widgets { get; set; }
        public int ReactionsCount { get; set; }
        public int RepliesCount { get; set; }
        public required int ChatId { get; set; }

        public virtual User? Author { get; set; }
        public virtual Chat? Chat { get; set; }
        public virtual ICollection<MessageReaction> Reactions { get; set; } = [];
        public virtual ICollection<User> Mentions { get; set; } = [];

        public MessageDTO ToDTO()
        {
            return new MessageDTO
            {
                Id = Id,
                Body = Body,
                Images = Images,
                Videos = Videos,
                Audios = Audios,
                PublishDate = PublishDate,
                ChildPostId = ChildPostId,
                MessageRepliedTo = MessageRepliedTo,
                AuthorId = AuthorId,
                Widgets = Widgets,
                ReactionsCount = ReactionsCount,
                RepliesCount = RepliesCount,
                ChatId = ChatId,
            };
        }
    }
}

File: universe/lib/models/data/message.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:universe/models/data/chat.dart';
import 'package:universe/models/data/user.dart';
import 'package:universe/models/data/widget.dart';
import 'package:uuid/uuid.dart';

part 'message.g.dart';

@JsonSerializable()
class Message {
  final int id;
  String? uid;
  final String body;
  final List<String> images;
  final List<String> videos;
  final List<String> audios;
  final DateTime publishDate;
  int? childPostId;
  int? messageRepliedTo;
  final String authorId;
  List<Widget>? widgets;
  int reactionsCount;
  int repliesCount;
  final int chatId;
  Chat? chat;
  User? author;

  Message({
    required this.id,
    this.uid,
    required this.body,
    required this.images,
    required this.videos,
    required this.audios,
    required this.publishDate,
    this.childPostId,
    this.messageRepliedTo,
    required this.authorId,
    this.widgets,
    required this.reactionsCount,
    required this.repliesCount,
    required this.chatId,
    this.chat,
    this.author,
  }) {
    uid ??= const Uuid().v4();
  }

  factory Message.fromJson(Map<String, dynamic> json) =>
      _$MessageFromJson(json);

  Map<String, dynamic> toJson() => _$MessageToJson(this);
}
